keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,420 @@
+
+# Proposal
+
+## User Stories
+
+#### Admin
+
+- As a cluster administrator, I can manage multiple object store providers via the Kubernetes API, so that I can see all object storage instances from a single pane, regardless of the backing storage vendor.
+
+#### User
+
+- As a developer, I can define my object storage needs in the same manifest as my workload, so that deployments are streamlined and encapsulated within the Kubernetes interface.
+- As a developer, I can define a manifest containing my workload and object storage configuration once, so that my app may be ported between clusters as long as the storage provided supports my designated data path protocol.
+- As a developer, I want to create a workload controller which is bucket API aware, so that it can dynamically connect workloads to object storage instances.
+
+## APIs
+
+> NOTE: For a graphical mapping of how these APIs relate to each other, see [API Relationships](#api-relationships).
+
+
+
+### Storage APIs
+
+---
+
+#### BucketRequest
+
+A namespaced API representing a workload’s need for a storage instance endpoint. Created in the workload's namespace.  A [BucketAccessRequest](#bucketaccessrequest) is required in order to gain credentialed access to the storage instance.
+
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  protocol: [3]
+  bucketPrefix: [4]
+  bucketClassName: [5]
+  secretName: [6]
+  bucketName: [7]
+status:
+  phase: [9]
+  conditions: 
+```
+
+1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer BucketRequest deletion until backend deletion ops succeed.
+1. `protocol`: specifies the desired protocol.  One of {“s3”, “gcs”, or “azureBlob”}.
+1. `bucketPrefix`: (Optional) prefix prepended to a randomly generated bucket name, eg. “yosemite-photos-". If empty no prefix is prepended.
+1. `bucketClassName`: Name of the target BucketClass.
+1. `secretName`: (optional) Secret in the BucketRequest's namespace storing credentials to be used by a workload for bucket access.
+1. `bucketName`: Name of the Bucket instance representing the desired storage instance endpoint.  Set by the COSI Controller for new Buckets.  Specified by the user when connecting to existing buckets.
+1. `phase`: 
+   - *Pending*: The controller has detected the new `Bucket` and begun provisioning operations
+   - *Bound*: Provisioning operations have completed and the `Bucket` has been bound to a `Bucket`.
```

Comment:
Good question.  We had planned on this being expanded as we prototype the automation.  We felt there's likely to be some intermittent phases that are worth communicating.

Reply:
When the delete case is considered and we are thinking of async deletes, we likely need a "DeletePending" phase. There has been some discussion about various "pending" phases, eg. "CreatePending", "GrantPending", "RevokePending". These phases would also be seen in the `Bucket` resource. The idea is that the admin and user can more easily see the current step should something get stuck...

Reply:
In case of errors, or stuck states, we should following the way of the pod statuses. Have a error status

- ErrBucketClassDoesNotSupportProtocol
- ErrBucketDeletionInProgress
etc. 

To answer @xing-yang's question. It might be better to keep them distinct states rather than use a boolean. it'll allow us to expand in the future to add more states. In case of the boolean, we will have a harder time adding new states without adding a new field, or breaking existing API.